packages/ai-chat/src/common/chat-agents.ts

```diff
@@ -389,28 +389,38 @@ export abstract class AbstractStreamParsingChatAgent extends AbstractChatAgent {
     }
 
     protected async addStreamResponse(languageModelResponse: LanguageModelStreamResponse, request: MutableChatRequestModel): Promise<void> {
-        for await (const token of languageModelResponse.stream) {
-            const newContents = this.parse(token, request);
-            if (isArray(newContents)) {
-                request.response.response.addContents(newContents);
-            } else {
-                request.response.response.addContent(newContents);
-            }
+        // Keep a buffer of the entire streamed text so far
+        let completeTextBuffer = '';
 
-            const lastContent = request.response.response.content.pop();
-            if (lastContent === undefined) {
-                return;
-            }
-            const text = lastContent.asString?.();
-            if (text === undefined) {
-                return;
-            }
+        // Create a marker content to identify the start of this stream
+        const streamMarkerContent = new MarkdownChatResponseContentImpl('');
+        request.response.response.addContent(streamMarkerContent);
+        const markerIndex = request.response.response.content.length - 1;
 
-            const result: ChatResponseContent[] = findFirstMatch(this.contentMatchers, text) ? this.parseContents(text, request) : [];
-            if (result.length > 0) {
-                request.response.response.addContents(result);
+        for await (const token of languageModelResponse.stream) {
+            // Process the token - get new content to add
+            const newContent = this.parse(token, request);
+
+            // If it's a text token, append to our buffer
+            if (isTextResponsePart(token) && token.content) {
+                completeTextBuffer += token.content;
+
+                // Parse the entire buffer to get the updated content including incomplete matches
+                const parsedContents = this.parseContents(completeTextBuffer, request);
+
+                // Get all content before our marker
+                const contentBeforeMarker = request.response.response.content.slice(0, markerIndex);
```

Comment:
I think the index needs to be recalculated in the else cases

Reply:
Great catch, I missed that! I reworked the algorithm. To keep it simple, we now only parse "uninterrupted" text. That means if there is a tool call in between, we won't retry parsing the text altogether again, but just keep what we've collected and parsed before.

Do you think this is a fair assumption? Thank you!